diff -up pdns-3.0-pre.20110327.2103/configure.orig pdns-3.0-pre.20110327.2103/configure
--- pdns-3.0-pre.20110327.2103/configure.orig	2011-03-28 09:44:22.690000393 +0200
+++ pdns-3.0-pre.20110327.2103/configure	2011-03-28 09:46:26.822000001 +0200
@@ -18224,7 +18224,7 @@ if test "${with_pgsql+set}" = set; then 
   withval=$with_pgsql; PGSQL_lib_check="$withval/lib/pgsql $with_pgsql/lib"
 	PGSQL_inc_check="$withval/include/pgsql"
 else
-  PGSQL_lib_check="/usr/local/pgsql/lib/pgsql /usr/local/lib/pgsql /opt/pgsql/lib/pgsql /usr/lib/pgsql /usr/local/pgsql/lib /usr/local/lib /opt/pgsql/lib /usr/lib"
+  PGSQL_lib_check="/usr/local/pgsql/lib/pgsql /usr/local/lib/pgsql /opt/pgsql/lib/pgsql /usr/lib/pgsql /usr/local/pgsql/lib /usr/local/lib /opt/pgsql/lib /usr/lib /usr/lib64"
 	PGSQL_inc_check="/usr/local/pgsql/include/pgsql /usr/include /usr/local/include/postgresql/ /usr/local/include /opt/pgsql/include/pgsql /opt/pgsql/include /usr/include/pgsql/ /usr/include/postgresql"
 fi
 
@@ -18244,8 +18244,7 @@ fi
 $as_echo_n "checking for PgSQL library directory... " >&6; }
 	PGSQL_libdir=
 	for m in $PGSQL_lib_check; do
-	        if test -d "$m" && \
-		   (test -f "$m/libpq.a" || test -f "$m/libpq++.a")
+	        if test -d "$m" && test -f "$m/libpq.so"
 	        then
 	                PGSQL_libdir=$m
 	                break
