--- pdns-3.1/modules/mongodbbackend/mongodbbackend.hh.orig	2012-05-04 12:13:23.000000000 +0200
+++ pdns-3.1/modules/mongodbbackend/mongodbbackend.hh	2012-09-18 13:09:19.000000000 +0200
@@ -1,13 +1,13 @@
 #ifndef MONGODB_HH
 #define MONGODB_HH
 
+#include "client/dbclient.h"
 #include "pdns/dnsbackend.hh"
 
 #undef VERSION
 #include <string>
 using std::string;
-
-#include "client/dbclient.h"
+using std::auto_ptr;
 
 class MONGODBException {
 public:
